### ISSUE TITLE
IObjectServiceDelegate are not composable

### ISSUE THREAD

**Issue by pcdavid** [eclipse-sirius/sirius-web#3036]
If two different extensions (E1 and E2) define their own custom `IObjectServiceDelegate`, they can not both work completely at the same time.

In `ComposedObjectService.getObject(IEditingContext, String)` we do:

```java
    @Override
    public Optional<Object> getObject(IEditingContext editingContext, String objectId) {
        var optionalDelegate = this.objectServiceDelegates.stream()
                .filter(delegate -> delegate.canHandle(editingContext))
                .findFirst();
        if (optionalDelegate.isPresent()) {
            return optionalDelegate.get().getObject(editingContext, objectId);
        }
        return this.defaultObjectService.getObject(editingContext, objectId);
    }
```

If E1 and E2 want to have any chance to get their `getObject(editingContext, objectId)` method called, they currently *must* return `true` to `canHandle(editingContext)`. However, when this is the case the "first" (on the order Spring will decide to inject them) which is found by `ComposedObjectService` will be the only one invoked. If it's E1, it can resolve its own objects and the defaults (by extending `DefaultObjectService`) but  will not know about potential other objects that E2 would be able to resolve, and E2 will never be called (until maybe the next reboot if it is injected before E1 this time).

One option would be for `IObjectServiceDelegate.canHandle(IEditingContext)` to also take the `String objectId` as it is the "guard method" for `getObject(IEditingContext editingContext, String objectId)` and `getContents(IEditingContext editingContext, String objectId)`. However this means that to answer this more precise question, `canHandle(IEditingContext, String)` some (most? all?) implementations will need to actully try to resolve the designated object just to return `true` and then be called again via `getObject(IEditingContext, String)` and re-do the same work. And a third and fourth time in the (relatively common) case where the next step will be to invoke `getContents()`:

1. canHandle(editinContext, objectId) => resolve the objectId, return true if found
2. getObject(editinContext, objectId) => resolve the objectId, this time to return it
3. canHandle(editinContext, objectId) => resolve the objectId, return true if found
4. getContents(editinContext, objectId) => resolve the objectId, this time to return its children

Another option would be to remove `IObjectServiceDelegate.canHandle(IEditingContext)`, leverage the fact that `IObjectServiceDelegate.getObject()` returns an optional and change `IObjectServiceDelegate.getContents()` to return an `Optional<List<Object>>` and return the first non-empty result from all the known delegates.

Another possible change (somewhat orthogonal to the above) would be to change `getContents(IEditingContext editinContext, String objectId)` into `getContents(IEditingContext editinContext, Object object)`. In the common case where we invoke `getContents` in a context where we have already resolved the objectId this would avoid some duplicate work. A default implementation could be added to `IObjectService[Delegate]` for compatibility:

```java
  // New method, takes an already resolved object
  List<Object> getContents(IEditingContext editingContext, Object object);

  // Old method, takes the id if that's all we have and resolve it first
  default List<Object> getContents(IEditingContext editingContext, String objectId) {
    return this.getObject(editingContext, String objectId).map(this::getContents).orElse(List.of());
  }
```

**Comment by pcdavid** [comment]
Some context: in my case I tried to implement an `RepresentationObjectServiceDelegate` that can see `IRepresentation`s as "semantic objects" (to work around #3023). For this particular case, maybe we would prefer to have "native" support for this in `DefaultObjectService`, but the issue would remain for other cases where multiple `IObjectServiceDelegate` would need to coexist.

**Comment by sbegaudeau** [comment]
- I think that the `canHandle` method should indeed take the `objectId`
- I think that the getContents method should only take the `Object object`, I'm not sure we need to give it back the editing context

I think I talked about this with Axel when we discussed future improvements in order to split the `IObjectService` interface into multiple smaller ones. I did started this work but it's currently a work in progress on a local branch on my machine. It is relevant for the extensibility of Sirius Web but it had a lower priority than other key architectural issues.

If you need as a quick fix, to had something like `objectId` to canHandle and change `getContents` to take the object just go for it, I'll make the split into smaller interfaces later. From my analysis the last time I worked on it, we could simply switch to `getContents(Object object)` without issues, unless I'm mistaken, just break the existing interface and change the signature, we will break it even further "soon" with the split.

**Comment by pcdavid** [comment]
For the moment I have hard-coded support for resolving `IRepresentation`s directly into `DefaultObjectService`. This is not ideal, but sufficient for me to move on with other issues revealed by #2890 on my own branch.
The new APIs that will replace `IObjectService` being developed in #3042 should solve this whole issue. Depending on what needs to be merged first, the changes proposed here to the current `IObjectService` may never need to be done.

**Comment by pcdavid** [comment]
I believe this can be closed; `IObjectServiceDelegate` no longer exists, and its finer-grained replacements like `IObjectSearchServiceDelegate` now have `boolean canHandle(IEditingContext editingContext, String objectId)`.
